### PR TITLE
allow to override the log level of stderr logging

### PR DESCRIPTION
**Pull request by leo-b**
PR against master for #3562

### CODE REVIEW

File: src/borg/testsuite/logger.py [hunk 1]
```diff
@@ -1,12 +1,27 @@
...
 from ..logger import find_parent_module, create_logger, setup_logging
+from .. import logger as _logger
 logger = create_logger()
 
 
+def reset_logging():
+    global logger
+    import importlib
+    # otherwise we'd need to reset the root logger and delete all other loggerss
+    importlib.reload(logging)
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/logger.py]: iirc, reload() has all sorts of issues. does it really work here?

typo: loggerss

File: src/borg/testsuite/logger.py [hunk 2]
```diff
@@ -16,11 +31,120 @@ def io_logger():
...
+# name: module, e.g. borg.archiver
+[formatter_stream]
+format=%%(levelname)s: %%(message)s
+
+[formatter_file]
+format=FILE %%(levelname)s: %%(message)s
+"""
+    logfile = tmpdir.join('borg.log')
+    # touch the file
+    logfile.open(mode='w')
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/logger.py]: this looks like truncate / create-new rather than touch.

File: src/borg/testsuite/logger.py [hunk 3]
```diff
@@ -39,7 +163,7 @@ def test_parent_module():
     assert find_parent_module() == __name__
 
 
-def test_lazy_logger():
+def test_lazy_logger(io_logger):
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/logger.py]: hm?

File: src/borg/testsuite/logger.py [hunk 4]
```diff
@@ -16,11 +31,120 @@ def io_logger():
...
+
+        @classmethod
+        def write_logging_conf(cls):
+            data = logging_conf % cls.cfg
+            cls.logging_config_file.write(data)
+
+        @classmethod
+        def logfile_contents(cls):
+            return open(cls.cfg['logfile_path']).read()
+    os.environ['BORG_LOGGING_CONF'] = str(context.logging_config_file)
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/logger.py]: an already present value will be lost

File: src/borg/testsuite/logger.py [hunk 5]
```diff
@@ -16,11 +31,120 @@ def io_logger():
...
+    yield context
+    del os.environ['BORG_LOGGING_CONF']
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/logger.py]: if meanwhile the key was already removed, this will crash

File: src/borg/testsuite/logger.py (hunk 1 above)
Comment (leo-b) [reply src/borg/testsuite/logger.py]: In the meantime I came to the conclusion that re-setting the logging config shouldn't be our method of choice. See my argumentation below...

File: src/borg/testsuite/logger.py (hunk 3 above)
Comment (leo-b) [reply src/borg/testsuite/logger.py]: All of borgs testing code relies on a global logging setup defined in conftest.py. But when testing the logging setup itself, it makes sense to reset it after every test function. My new teardown_function() does that.
Thus ``test_lazy_logger()`` doesn't have a logging setup anymore.
An alternative would be to call ``setup_logging()`` at the beginning.

**Comment by leo-b** [comment]
Resetting the initial state of pythons logging is really hard. Removing all loggers isn't enough. A reference to the root logger is kept in several places inside the python logging module and the whole hierarchic logging configuration is kept inside the logging module.

All borg tests rely on a logging configuration set up in conftest.py. Other previously executed borg test modules have already initialized most of borgs modules, which themselves keep their own reference to their module level logger.

That's why I just came to the conclusion that destroying and re-building the existing logging config won't work.

However, testing the logging setup itself and especially testing a file based logging config requires a fresh environment. I am not a pytest expert... Is there a way to force running the logging tests in separate processes for proper isolation? I guess pytest-xdist would be the way to go...

**Comment by ThomasWaldmann** [comment]
pytest-xdist is usually used for parallelization. but i don't think we should require it for correct execution of tests.

**Comment by leo-b** [comment]
Hmm..
No [py.test --boxed](https://github.com/pytest-dev/pytest-xdist#running-tests-in-a-boxed-subprocess)?
Any other idea?
I think the only way to isolate logging setup properly is to run those tests in separate processes.
If not, the tests could always interfere with each other or damage the global logging setup that all other tests rely on.
Unfortunately I don't think I'll be able to contribute an implementation for that...

**Comment by ThomasWaldmann** [comment]
Well, not really. Except the usual "if something can't be done in a reasonable way, don't do it".

**Comment by leo-b** [comment]
Eventually I digged into pytest and managed to find a way to run particular tests in a separate forked process.
This is based on the [_pytest-forked_ plugin](https://github.com/pytest-dev/pytest-forked), which is also required by _pytest-xdist_, so it shouldn't add any new requirements.
If _pytest-xdist_ isn't available, the corresponding tests are simply skipped with an appropriate message.

**Pull request closed without merging**